kernel/src/process/rlimit.rs

```diff
@@ -112,55 +108,70 @@ pub struct RawRLimit64 {
-            return_errno_with_message!(Errno::EINVAL, "invalid rlimit");
-        }
-        Ok(Self {
-            cur: AtomicU64::new(cur_),
-            max: AtomicU64::new(max_),
+    pub(self) const fn new(cur: u64, max: u64) -> Self {
+        assert!(cur <= max, "the current rlimit exceeds the max rlimit");
+        Self {
+            cur: AtomicU64::new(cur),
+            max: AtomicU64::new(max),
             lock: SpinLock::new(()),
-        })
+        }
     }
 
-    /// Gets the current rlimit without synchronization.
+    /// Returns the current rlimit without synchronization.
     pub fn get_cur(&self) -> u64 {
         self.cur.load(Ordering::Relaxed)
     }
 
-    /// Gets the max rlimit without synchronization.
-    pub fn get_max(&self) -> u64 {
+    /// Returns the max rlimit without synchronization.
+    fn get_max(&self) -> u64 {
         self.max.load(Ordering::Relaxed)
     }
 
-    /// Gets the rlimit with synchronization.
+    /// Returns the rlimit with synchronization.
     ///
     /// Only called when handling the `getrlimit` or `prlimit` syscall.
-    pub fn get_cur_and_max(&self) -> (u64, u64) {
+    pub fn get_raw_rlimit(&self) -> RawRLimit64 {
         let _guard = self.lock.lock();
-        (self.get_cur(), self.get_max())
+        RawRLimit64 {
+            cur: self.cur.load(Ordering::Relaxed),
+            max: self.max.load(Ordering::Relaxed),
+        }
     }
 
-    /// Sets the rlimit with synchronization.
+    /// Sets the rlimit with synchronization and returns the old value.
     ///
-    /// Only called when handling the `setrlimit` or `prlimit` syscall
-    /// or during init process creation.
-    pub fn set_cur_and_max(&self, new_cur: u64, new_max: u64) -> Result<()> {
-        if new_cur > new_max {
-            return_errno_with_message!(Errno::EINVAL, "invalid rlimit");
+    /// Only called when handling the `setrlimit` or `prlimit` syscall.
+    pub fn set_raw_rlimit(&self, new: RawRLimit64, ctx: &Context) -> Result<RawRLimit64> {
+        if new.cur > new.max {
+            return_errno_with_message!(Errno::EINVAL, "the current rlimit exceeds the max rlimit");
         }
         let _guard = self.lock.lock();
-        self.cur.store(new_cur, Ordering::Relaxed);
-        self.max.store(new_max, Ordering::Relaxed);
-        Ok(())
+        if new.max > self.get_max() {
+            let init_user_ns = UserNamespace::get_init_singleton();
+            init_user_ns.check_cap(CapSet::SYS_RESOURCE, ctx.posix_thread)?;
+        }
+        let old = RawRLimit64 {
+            cur: self.cur.load(Ordering::Relaxed),
+            max: self.max.load(Ordering::Relaxed),
+        };
+        self.set_raw_rlimit_unchecked(new);
+        Ok(old)
+    }
+
+    /// Sets the rlimit _without_ synchronization and permission check.
+    ///
+    /// Only called during init process creation.
+    pub(self) fn set_raw_rlimit_unchecked(&self, new: RawRLimit64) {
+        assert!(new.cur < new.max);
```

Comment:
`cur < max`? Will it work?
```suggestion
        assert!(cur <= max, "the current rlimit exceeds the max rlimit");
```